### ISSUE TITLE
I have a llama2-7b model and a checkpoint fine-tuned using p-tuning. How do I load the base model and checkpoint using PEFT for inference?

### ISSUE THREAD

**Issue by linguoqi** [huggingface/peft#1500]
### System Info

unbuntu:5.15.0-92-generic
python=3.10
peft =0.8.2

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

```python
import os
os.environ['CUDA_VISIBLE_DEVICES'] = '7'
from peft import PromptEncoder, PromptEncoderConfig
from transformers import LlamaForCausalLM, AutoTokenizer , TrainingArguments ,Trainer , default_data_collator
from transformers import AutoModelForSeq2SeqLM
from peft import get_peft_model
from modelscope import MsDataset
import conf as co




peft_config = PromptEncoderConfig(
    peft_type="P_TUNING",
    task_type='CAUSAL_LM',
    num_virtual_tokens=5,
    encoder_dropout = 0.3,
    # token_dim=768,
    num_transformer_submodules=1,
    num_attention_heads=2,
    num_layers=1,
    # encoder_reparameterization_type="MLP",
    encoder_hidden_size=16,
)

model = LlamaForCausalLM.from_pretrained("/tf/model/Llama-2-7b-chat-ms/")


model = get_peft_model(model, peft_config)
model.print_trainable_parameters()
tokenizer = AutoTokenizer.from_pretrained("/tf/model/Llama-2-7b-chat-ms/")



train_dataset = MsDataset.load(dataset_name = '/tf/user/lgq/data',data_files={'train': co.train_data_path}).to_hf_dataset()
val_dataset = MsDataset.load(dataset_name = '/tf/user/lgq/data',data_files={'val': co.val_data_path}).to_hf_dataset().select(range(4))


def tokenize_function(examples):
    return tokenizer(examples['instruction'], examples['input'],examples['output'], padding='max_length', truncation=True, max_length=512)

train_dataset = train_dataset.map(tokenize_function, batched=True)
val_dataset = val_dataset.map(tokenize_function, batched=True)


training_args = TrainingArguments(
    output_dir= co.output_dir,
    per_device_train_batch_size=4,
    learning_rate=4.5e-07,
    num_train_epochs=2,
    eval_steps=1000
)



trainer = Trainer(model, training_args, train_dataset=train_dataset,
                    eval_dataset=val_dataset,
                    data_collator=default_data_collator,)
trainer.train()
```

### Expected behavior

I have a llama2-7b model and a checkpoint that has been fine-tuned using p-tuning. How can I load the base model and checkpoint using PEFT for inference? I tried to find this method in the PEFT GitHub, but I couldn't find it. Thank you very much for your help.

**Comment by BenjaminBossan** [comment]
Does [this section](https://huggingface.co/docs/peft/v0.8.2/en/quicktour#inference) help? Of course, you need to save the model with `save_pretrained` first after you finish training.

**Comment by linguoqi** [comment]
This method can load the checkpoint. I used the trainer from transformers for model fine-tuning, but it seems that this method does not support saving the pre-trained model, only the checkpoint?

PeftModel.from_pretrained(model, peft_model_id)

**Comment by BenjaminBossan** [comment]
> but it seems that this method does not support saving the pre-trained model, only the checkpoint?

Do I understand correctly that you want to save the whole model, i.e. base model + adapter, instead of just the adapter? In general, this shouldn't be necessary, as you can load the base model like you normally would and add the adapter on top using `PeftModel.from_pretrained(model, peft_model_id)`. However, if you really need to save the whole model as a single checkpoint, you can call `model = model.merge_and_unload(); model.save_pretrained(...)`, which gives you a single (big) file.

**Comment by linguoqi** [comment]
Thank you for your reply. Additionally, the following error occurred while performing inference with a model fine-tuned using PEFT.
![企业微信截图_1708742615384](https://github.com/huggingface/peft/assets/157370038/c0c43b98-dd8a-4cd4-9adb-6f4300d9dd97)


My training code is as follows:
`import os
os.environ['CUDA_VISIBLE_DEVICES'] = '5'
os.environ['PYTORCH_CUDA_ALLOC_CONF'] = 'expandable_segments:True'
from peft import PromptEncoder, PromptEncoderConfig
from transformers import LlamaForCausalLM, AutoTokenizer , TrainingArguments ,Trainer , default_data_collator
from transformers import AutoModelForSeq2SeqLM
from peft import get_peft_model
from modelscope import MsDataset
import conf as co
import torch
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")



peft_config = PromptEncoderConfig(
    peft_type="P_TUNING",
    task_type='CAUSAL_LM',
    num_virtual_tokens=1,
    encoder_dropout = 0,
    # token_dim=768,
    num_transformer_submodules=1,
    num_attention_heads=1,
    num_layers=1,
    # encoder_reparameterization_type="MLP",
    encoder_hidden_size=2,
)

model = LlamaForCausalLM.from_pretrained("/tf/model/Llama-2-7b-chat-ms/")


model = get_peft_model(model, peft_config)
model.print_trainable_parameters()
tokenizer = AutoTokenizer.from_pretrained("/tf/model/Llama-2-7b-chat-ms/")



train_dataset = MsDataset.load(dataset_name = '/tf/user/lgq/data',data_files={'train': co.train_data_path}).to_hf_dataset()
val_dataset = MsDataset.load(dataset_name = '/tf/user/lgq/data',data_files={'val': co.val_data_path}).to_hf_dataset().select(range(4))


def tokenize_function(examples):
    return tokenizer(examples['instruction'], examples['input'],examples['output'], padding='max_length', truncation=True, max_length=1024)

train_dataset = train_dataset.map(tokenize_function, batched=True)
val_dataset = val_dataset.map(tokenize_function, batched=True)


training_args = TrainingArguments(
    output_dir= './peft_model_card_5',
    per_device_train_batch_size=1,
    learning_rate=4.5e-07,
    num_train_epochs=4,
    eval_steps=1000,
    # save_steps=21,
    # max_steps=20
)`

My inference code is as follows:
`import os
os.environ['CUDA_VISIBLE_DEVICES'] = '0,1'
# os.environ['PYTORCH_CUDA_ALLOC_CONF'] = 'expandable_segments:True'
from peft import PromptEncoder, PromptEncoderConfig
from transformers import LlamaForCausalLM, AutoTokenizer , TrainingArguments ,Trainer , default_data_collator
from transformers import AutoModelForSeq2SeqLM
from peft import get_peft_model, PeftModel
from modelscope import MsDataset
import conf as co
import torch
# device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
import gradio as gr
device = 'cuda'

peft_config = PromptEncoderConfig(
    peft_type="P_TUNING",
    task_type='CAUSAL_LM',
    num_virtual_tokens=10,
    encoder_dropout = 0,
    # token_dim=768,
    num_transformer_submodules=1,
    num_attention_heads=12,
    num_layers=6,
    # encoder_reparameterization_type="MLP",
    encoder_hidden_size=8,
)

model = LlamaForCausalLM.from_pretrained("/tf/model/Llama-2-7b-chat-ms/", device_map='auto')


model = get_peft_model(model, peft_config)
model.print_trainable_parameters()
tokenizer = AutoTokenizer.from_pretrained("/tf/model/Llama-2-7b-chat-ms/")
PeftModel.from_pretrained(model, '/tf/user/lgq/Power_Dispatch_LLM/train/peft_model/checkpoint-6000')




def gradio_chat_demo():
    def model_chat(query):
        print(query)
        # res = model(**tokenizer(query, return_tensors='pt'))
        inputs = tokenizer(query, return_tensors="pt")
        inputs.to(device)
        # print(inputs)
        generate_ids = model.generate(inputs.input_ids, max_length=1024)
        res = str(tokenizer.batch_decode(generate_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False)[0])
        return res
    demo = gr.Interface(fn=model_chat, inputs="text", outputs="text")
    demo.queue().launch(root_path="/7b-app", height=1000, share=False, server_port=8081, server_name='0.0.0.0')
    # demo.queue().launch(share=False, server_port=8080, server_name='0.0.0.0')
    return demo


demo = gradio_chat_demo()`

**Comment by BenjaminBossan** [comment]
For inference, you should not call `model = get_peft_model(model, peft_config)`, `PeftModel.from_pretrained` with the base model should be enough. Could you please try if that fixes your issue?

Also, please, when posting error messages, please paste the whole error message as text, not a screenshot.
